src/install/install.zig

```diff
@@ -10089,162 +10120,606 @@ pub const PackageManager = struct {
+            bun.OSPathLiteral("node_modules"),
+            bun.OSPathLiteral(".git"),
+            bun.OSPathLiteral("CMakeFiles"),
+        };
+
+        const FileCopier = struct {
+            pub fn copy(
+                destination_dir_: std.fs.Dir,
+                walker: *Walker,
+            ) !u32 {
+                var real_file_count: u32 = 0;
+
+                var copy_file_state: bun.CopyFileState = .{};
+                var pathbuf: bun.PathBuffer = undefined;
+                var pathbuf2: bun.PathBuffer = undefined;
+                // _ = pathbuf; // autofix
+
+                while (try walker.next()) |entry| {
+                    if (entry.kind != .file) continue;
+                    real_file_count += 1;
+                    const openFile = std.fs.Dir.openFile;
+                    const createFile = std.fs.Dir.createFile;
+
+                    // - rename node_modules/$PKG/$FILE -> node_modules/$PKG/$TMPNAME
+                    // - create node_modules/$PKG/$FILE
+                    // - copy: cache/$PKG/$FILE -> node_modules/$PKG/$FILE
+                    // - unlink: $TMPDIR/$FILE
+                    if (comptime bun.Environment.isWindows) {
+                        var tmpbuf: [1024]u8 = undefined;
+                        const basename = bun.strings.fromWPath(pathbuf2[0..], entry.basename);
+                        const tmpname = bun.span(bun.fs.FileSystem.instance.tmpname(basename, tmpbuf[0..], bun.fastRandom()) catch |e| {
+                            Output.prettyError("<r><red>error<r>: copying file {s}", .{@errorName(e)});
+                            Global.crash();
+                        });
+
+                        const entrypath = bun.strings.fromWPath(pathbuf[0..], entry.path);
+                        pathbuf[entrypath.len] = 0;
+                        const entrypathZ = pathbuf[0..entrypath.len :0];
+
+                        if (bun.sys.renameatConcurrently(
+                            bun.toFD(destination_dir_.fd),
+                            entrypathZ,
+                            bun.toFD(destination_dir_.fd),
+                            tmpname,
+                        ).asErr()) |e| {
+                            Output.prettyError("<r><red>error<r>: copying file {}", .{e});
+                            Global.crash();
+                        }
+
+                        var in_file = openFile(entry.dir, basename, .{ .mode = .read_only }) catch |e| {
+                            Output.prettyError("<r><red>error<r>: opening node_modules file {s} {s} ()", .{ @errorName(e), entrypath });
+                            Global.crash();
+                        };
+                        defer in_file.close();
+
+                        const mode = in_file.mode() catch @panic("OH NO");
+                        var outfile = createFile(destination_dir_, entrypath, .{ .mode = mode }) catch @panic("OH NO");
```

Comment:
This will cause AV & other software currently accessing the file to error

Instead, we should close the file immediately before we begin copying.